File: packages/sdk-node/src/credentials.ts
import type { HttpClient } from "./http.js";
import type { CredentialRecord, ServiceId } from "./types.js";

/**
 * Credential API: attach / rotate / revoke real upstream keys held by the
 * control plane. These keys never enter a sandbox; they're only used by the
 * phantom proxy to re-inject auth headers.
 */
export class Credentials {
  constructor(private readonly http: HttpClient) {}

  /** List all configured credentials (metadata only — no secrets). */
  async list(): Promise<CredentialRecord[]> {
    return this.http.request<CredentialRecord[]>({
      method: "GET",
      path: "/v1/credentials",
    });
  }

  /** Attach a new credential or rotate an existing one. */
  async put(params: {
    service: ServiceId;
    secret: string;
  }): Promise<CredentialRecord> {
    return this.http.request<CredentialRecord>({
      method: "POST",
      path: "/v1/credentials",
      body: params,
    });
  }

  /** Remove a credential. Instantly invalidates associated phantom tokens. */
  async delete(service: ServiceId): Promise<void> {
    await this.http.request<void>({
      method: "DELETE",
      path: `/v1/credentials/${encodeURIComponent(service)}`,
    });
  }
}

File: packages/sdk-python/src/agentjail/credentials.py
"""Credentials — put/list/delete provider API keys."""

from __future__ import annotations

from ._http import HttpClient
from .types import CredentialRecord, ServiceId


class Credentials:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def put(self, *, service: ServiceId, secret: str) -> CredentialRecord:
        return self._http.request(
            "POST", "/v1/credentials", json={"service": service, "secret": secret}
        )

    def list(self) -> list[CredentialRecord]:
        return self._http.request("GET", "/v1/credentials")

    def delete(self, service: ServiceId) -> None:
        self._http.request("DELETE", f"/v1/credentials/{service}")
